insuree/schema.py

```diff
@@ -453,3 +458,34 @@ def _get_additional_filter(sender, additional_filter, user, signal: Signal):
         )
         filters_from_signal = _read_signal_results(results_signal)
     return filters_from_signal
+
+def _get_contextual_insuree_officers(info, **kwargs):
+        if not info.context.user.has_perms(InsureeConfig.gql_query_insuree_officers_perms):
+            raise PermissionDenied(_("unauthorized"))
+
+        user = info.context.user
+        i_user = getattr(user, '_u', None)
+
+        if i_user:
+            user_roles = UserRole.objects.filter(user_id=i_user.id, validity_to__isnull=True)
+            roles = list(
+                Role.objects.filter(
+                    id__in=user_roles.values_list("role_id", flat=True),
+                    validity_to__isnull=True
+                ).values_list("name", flat=True)
+            )
+            # If the user is an Enrolment Officer (EO)
+            if "Enrolment Officer" in roles:
+                return Officer.objects.filter(id=user.officer.id, validity_to__isnull=True)
+
+            # Non-EO user
+            user_districts = UserDistrict.get_user_districts(i_user)
+            if user_districts:
+                location_uuids = [d.location.uuid for d in user_districts if d.location]
+                officers = Officer.objects.filter(location__uuid__in=location_uuids, validity_to__isnull=True).distinct()
+                if officers.exists():
```

Comment:
to avoid database effort, use location_id which prevent a join in the query AND in the filter
```suggestion
            if user_districts:
                location_ids = [d.location_id for d in user_districts if d.location]
                officers = Officer.objects.filter(location__id__in=location_ids, validity_to__isnull=True).distinct()
                if officers.exists():
```